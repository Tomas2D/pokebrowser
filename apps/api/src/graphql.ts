import {
  NexusGenFieldTypes,
  NexusGenInputs,
} from "./graphql/generated/typings";

export type {
  NexusGenObjects,
  NexusGenInputs,
} from "./graphql/generated/typings";

export type Pokemon = Required<NexusGenFieldTypes["Pokemon"]>;
export type PokemonMeta = Required<NexusGenFieldTypes["ListPokemonMeta"]>;
export type User = Required<NexusGenFieldTypes["User"]>;
export type Attack = NexusGenFieldTypes["Attack"];
export type EvolutionRequirement = NexusGenFieldTypes["EvolutionRequirement"];
export type Height = NexusGenFieldTypes["Height"];
export type Weight = NexusGenFieldTypes["Weight"];
export type Type = NexusGenFieldTypes["PokemonType"];
export type PokemonFilters = NexusGenInputs["ListPokemonFilters"];
