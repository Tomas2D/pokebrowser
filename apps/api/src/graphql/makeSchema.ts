import { QueryResolver } from "@app/graphql/resolver/QueryResolver";
import { MutationResolver } from "@app/graphql/resolver/MutationResolver";
import { connectionPlugin, makeSchema as graphqlMakeSchema } from "nexus";
import * as schemaTypes from "./schema/types";

export function makeSchema() {
  return graphqlMakeSchema({
    plugins: [connectionPlugin()],
    types: [Object.values(schemaTypes), QueryResolver, MutationResolver],
    nonNullDefaults: {
      input: true,
      output: true,
    },
    outputs: {
      schema: __dirname + "/generated/schema.graphql",
      typegen: __dirname + "/generated/typings.ts",
    },
  });
}
