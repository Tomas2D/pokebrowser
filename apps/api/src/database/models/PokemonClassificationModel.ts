import { Model } from "objection";
import { PokemonModel } from "./PokemonModel";

export class PokemonClassificationModel extends Model {
  id!: number;
  name!: string;

  static get tableName() {
    return "pokemon_classification";
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["name"],
      properties: {
        id: { type: "integer" },
        name: { type: "string", minLength: 1, maxLength: 255 },
      },
    };
  }

  static get relationMappings() {
    return {
      pokemons: {
        relation: Model.HasManyRelation,
        modelClass: PokemonModel,
        join: {
          from: "pokemon_classification.id",
          to: "pokemon.classificationId",
        },
      },
    };
  }
}
