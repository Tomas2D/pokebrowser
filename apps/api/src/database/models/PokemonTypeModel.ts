import { Model } from "objection";
import { PokemonModel } from "./PokemonModel";

export class PokemonTypeModel extends Model {
  id!: number;
  name!: string;

  static get tableName() {
    return "pokemon_type";
  }

  static get relationMappings() {
    return {
      pokemon: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonModel,
        join: {
          from: "pokemon_type.id",
          through: {
            from: "pokemon_type_pokemon.typeId",
            to: "pokemon_type_pokemon.pokemonId",
          },
          to: "pokemon.id",
        },
      },
    };
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["name"],
      properties: {
        id: { type: "integer" },
        name: { type: "string", minLength: 1, maxLength: 255 },
      },
    };
  }
}
