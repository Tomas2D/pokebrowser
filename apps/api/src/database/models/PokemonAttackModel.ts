import { Model } from "objection";
import { PokemonModel } from "./PokemonModel";
import { PokemonTypeModel } from "@app/database/models/PokemonTypeModel";

export class PokemonAttackModel extends Model {
  id!: number;
  name!: string;
  type!: PokemonTypeModel;
  damage!: number;
  category!: "fast" | "special";

  static get tableName() {
    return "pokemon_attack";
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["name", "type", "damage", "category"],
      properties: {
        id: { type: "integer" },
        name: { type: "string", minLength: 1, maxLength: 255 },
        category: { type: "string", minLength: 1, maxLength: 255 },
        damage: { type: "integer" },
      },
    };
  }

  static get relationMappings() {
    return {
      pokemons: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonModel,
        join: {
          from: "pokemon_attack.id",
          through: {
            from: "pokemon_attack_pokemon.attackId",
            to: "pokemon_attack_pokemon.pokemonId",
          },
          to: "pokemon.id",
        },
      },
      type: {
        relation: Model.BelongsToOneRelation,
        modelClass: PokemonTypeModel,
        join: {
          from: `pokemon_attack.typeId`,
          to: `pokemon_type.id`,
        },
      },
    };
  }
}
