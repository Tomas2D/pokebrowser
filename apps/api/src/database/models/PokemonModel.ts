import { Model } from "objection";
import { PokemonAttackModel } from "./PokemonAttackModel";
import { PokemonTypeModel } from "./PokemonTypeModel";
import { PokemonClassificationModel } from "@app/database/models/PokemonClassificationModel";
import { PokemonVoteModel } from "@app/database/models/PokemonVoteModel";
import { UserModel } from "@app/database/models/UserModel";

export const PokemonRelations = {
  VOTES: "votes",
  ATTACKS: "attacks",
  PREVIOUS_EVOLUTION: "previousEvolution",
  NEXT_EVOLUTION: "nextEvolution",
  WEAKNESSES: "weaknesses",
  RESISTANTS: "resistants",
  CLASSIFICATION: "classification",
  TYPES: "types",
};

import cursor from "objection-cursor";
const cursorMixin = cursor({
  nodes: true,
  pageInfo: {
    hasNext: true,
    hasPrevious: true,
  },
});

export class PokemonModel extends cursorMixin(Model) {
  id!: number;
  name!: string;
  slug!: string;
  classification!: PokemonClassificationModel;
  commonCaptureArea!: string | null;
  evolutionRequirementAmount!: number | null;
  evolutionRequirementName!: string | null;
  weightMaximum!: number;
  weightMinimum!: number;
  heightMaximum!: number;
  heightMinimum!: number;
  fleeRate!: number;
  maxCp!: number;
  maxHp!: number;
  types!: PokemonTypeModel[];
  resistants!: PokemonTypeModel[];
  weaknesses!: PokemonTypeModel[];
  attacks!: PokemonAttackModel[];
  nextEvolution!: PokemonModel | null;
  previousEvolution!: PokemonModel | null;
  votes!: PokemonVoteModel[];

  static get idColumn() {
    return "id";
  }
  static get tableName() {
    return "pokemon";
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["name", "slug"],
      properties: {
        id: { type: "number" },
        name: { type: "string" },
        slug: { type: "string" },
        evolutionRequirementAmount: { type: ["number", "null"] },
        evolutionRequirementName: { type: ["string", "null"] },
        commonCaptureArea: { type: ["string", "null"] },
        weightMinimum: { type: "number" },
        weightMaximum: { type: "number" },
        heightMinimum: { type: "number" },
        heightMaximum: { type: "number" },
        fleeRate: { type: "number" },
        maxCp: { type: "number" },
        maxHp: { type: "number" },
      },
    };
  }

  // @ts-expect-error recursive circular dependency due to "self-to-self" relations
  static get relationMappings() {
    return {
      [PokemonRelations.TYPES]: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonTypeModel,
        join: {
          from: "pokemon.id",
          through: {
            from: "pokemon_type_pokemon.pokemonId",
            to: "pokemon_type_pokemon.typeId",
          },
          to: "pokemon_type.id",
          orderBy: "pokemon_type.id",
        },
      },
      [PokemonRelations.RESISTANTS]: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonTypeModel,
        join: {
          from: "pokemon.id",
          through: {
            from: "pokemon_resistant_pokemon.pokemonId",
            to: "pokemon_resistant_pokemon.typeId",
          },
          to: "pokemon_type.id",
          orderBy: "pokemon_type.id",
        },
      },
      [PokemonRelations.WEAKNESSES]: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonTypeModel,
        join: {
          from: "pokemon.id",
          through: {
            from: "pokemon_weakness_pokemon.pokemonId",
            to: "pokemon_weakness_pokemon.typeId",
          },
          to: "pokemon_type.id",
          orderBy: "pokemon_type.id",
        },
      },
      [PokemonRelations.CLASSIFICATION]: {
        relation: Model.BelongsToOneRelation,
        modelClass: PokemonClassificationModel,
        join: {
          from: `pokemon.classificationId`,
          to: `pokemon_classification.id`,
          orderBy: "pokemon_classification.id",
        },
      },
      [PokemonRelations.ATTACKS]: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonAttackModel,
        join: {
          from: "pokemon.id",
          through: {
            from: "pokemon_attack_pokemon.pokemonId",
            to: "pokemon_attack_pokemon.attackId",
          },
          to: "pokemon_attack.id",
          orderBy: "pokemon_attack.id",
        },
      },
      [PokemonRelations.PREVIOUS_EVOLUTION]: {
        relation: Model.BelongsToOneRelation,
        modelClass: PokemonModel,
        join: {
          from: "pokemon.id",
          to: "pokemon.nextEvolutionId",
        },
      },
      [PokemonRelations.NEXT_EVOLUTION]: {
        relation: Model.BelongsToOneRelation,
        modelClass: PokemonModel,
        join: {
          from: "pokemon.nextEvolutionId",
          to: "pokemon.id",
        },
      },
      [PokemonRelations.VOTES]: {
        relation: Model.ManyToManyRelation,
        modelClass: UserModel,
        join: {
          from: "pokemon.id",
          through: {
            from: "pokemon_vote.pokemonId",
            to: "pokemon_vote.userId",
          },
          to: "user.id",
          orderBy: "user.id",
        },
      },
    };
  }
}
