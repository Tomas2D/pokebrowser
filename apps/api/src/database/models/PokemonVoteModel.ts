import { Model } from "objection";
import { UserModel } from "@app/database/models/UserModel";
import { PokemonModel } from "@app/database/models/PokemonModel";

export class PokemonVoteModel extends Model {
  userId!: string;
  pokemonId!: number;

  static get idColumn() {
    return "id";
  }
  static get tableName() {
    return "pokemon_vote";
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["userId", "pokemonId"],
      properties: {
        userId: { type: "string" },
        pokemonId: { type: "integer" },
      },
    };
  }

  static get relationMappings() {
    return {
      user: {
        relation: Model.BelongsToOneRelation,
        modelClass: UserModel,
        join: {
          from: "pokemon_vote.userId",
          to: "user.id",
        },
      },
      pokemon: {
        relation: Model.BelongsToOneRelation,
        modelClass: PokemonModel,
        join: {
          from: "pokemon_vote.pokemonId",
          to: "pokemon.id",
        },
      },
    };
  }
}
