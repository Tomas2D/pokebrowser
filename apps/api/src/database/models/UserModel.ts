import { Model } from "objection";
import { PokemonModel } from "@app/database/models/PokemonModel";
import { PokemonVoteModel } from "@app/database/models/PokemonVoteModel";

export class UserModel extends Model {
  id!: string;
  createdAt!: Date;

  static get idColumn() {
    return "id";
  }

  static get tableName() {
    return "user";
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["id"],
      properties: {
        id: { type: "string" },
        createdAt: { type: "integer" },
      },
    };
  }

  static get relationMappings() {
    return {
      voteIds: {
        relation: Model.HasManyRelation,
        modelClass: PokemonVoteModel,
        join: {
          from: "user.id",
          to: "pokemon_vote.userId",
        },
      },
      pokemons: {
        relation: Model.ManyToManyRelation,
        modelClass: PokemonModel,
        join: {
          from: "user.id",
          through: {
            from: "pokemon_vote.userId",
            to: "pokemon_vote.pokemonId",
          },
          to: "pokemon.id",
        },
      },
    };
  }
}
