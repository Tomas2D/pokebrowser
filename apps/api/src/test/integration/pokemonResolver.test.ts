import { parse } from "graphql";
import { assertSingleValue, sortById } from "@app/test/utils/graphql";
import { destroyTestApp, initTestApp, TestApp } from "@app/test/utils/app";
import { PokemonFilters } from "@app/graphql";
import { PokemonModel } from "@app/database/models/PokemonModel";

describe("Pokemon", async () => {
  let app: TestApp;

  beforeAll(async () => {
    app = await initTestApp();
  });

  afterAll(async () => {
    await destroyTestApp(app);
  });

  it("returns list of pokemons", async () => {
    const result = await app.executor({
      variables: {
        filters: {
          size: 2,
          offset: 0,
        } as PokemonFilters,
      },
      document: parse(/* GraphQL */ `
        query listPokemon($filters: ListPokemonFilters!) {
          listPokemon(filters: $filters) {
            meta {
              total
              hasMore
            }
            edges {
              id
              name
              slug
              types {
                id
                name
              }
              maxHp
              maxCp
            }
          }
        }
      `),
    });

    assertSingleValue(result);

    expect(result.errors).toBeUndefined();
    expect(result.data).toBeDefined();
    expect(result.data.listPokemon.edges).toMatchInlineSnapshot(`
      [
        {
          "id": 1,
          "maxCp": 951,
          "maxHp": 1071,
          "name": "Bulbasaur",
          "slug": "bulbasaur",
          "types": [
            {
              "id": 1,
              "name": "Grass",
            },
            {
              "id": 2,
              "name": "Poison",
            },
          ],
        },
        {
          "id": 2,
          "maxCp": 1483,
          "maxHp": 1632,
          "name": "Ivysaur",
          "slug": "ivysaur",
          "types": [
            {
              "id": 1,
              "name": "Grass",
            },
            {
              "id": 2,
              "name": "Poison",
            },
          ],
        },
      ]
    `);
    expect(result.data.listPokemon.meta).toMatchInlineSnapshot(`
      {
        "hasMore": true,
        "total": 30,
      }
    `);
  });

  it("paginates correctly", async () => {
    const runQuery = async (offset: number) => {
      const result = await app.executor({
        variables: {
          filters: {
            size: 5,
            offset,
          } as PokemonFilters,
        },
        document: parse(/* GraphQL */ `
          query listPokemon($filters: ListPokemonFilters!) {
            listPokemon(filters: $filters) {
              meta {
                total
                hasMore
              }
              edges {
                id
                types {
                  id
                  name
                }
                attacks {
                  id
                  name
                  damage
                }
                nextEvolution {
                  id
                }
                previousEvolution {
                  id
                }
              }
            }
          }
        `),
      });

      assertSingleValue(result);

      return result;
    };

    let ids = new Set();
    let totalCount = Infinity;
    let offset = 0;

    while (ids.size < totalCount) {
      const { errors, data } = await runQuery(offset);
      expect(errors).toBeUndefined();
      expect(data).toBeDefined();

      if (totalCount === Infinity) {
        totalCount = data.listPokemon.meta.total;
      }
      expect(totalCount).toBe(data.listPokemon.meta.total);

      for (const { id } of data.listPokemon.edges) {
        expect(ids.has(id)).toBe(false);
        ids.add(id);
        offset++;
      }

      if (ids.size < totalCount) {
        expect(data.listPokemon.meta.hasMore).toBe(true);
      } else {
        expect(data.listPokemon.meta.hasMore).toBe(false);
      }
    }
  });

  it("filters pokemons by type", async () => {
    const type = { id: 1, name: "Electric" };
    const size = 3;

    const result = await app.executor({
      variables: {
        filters: {
          size,
          typeId: [type.id],
        } as PokemonFilters,
      },
      document: parse(/* GraphQL */ `
        query listPokemon($filters: ListPokemonFilters!) {
          listPokemon(filters: $filters) {
            edges {
              id
              types {
                id
                name
              }
            }
          }
        }
      `),
    });

    assertSingleValue(result);

    const { data, errors } = result;

    expect(errors).toBeUndefined();
    expect(data).toBeDefined();

    expect(data.listPokemon.edges).toHaveLength(size);
    expect(
      data.listPokemon.edges.every((edge: any) =>
        edge.types.some(({ id }: typeof edge) => id === type.id)
      )
    ).toBe(true);
  });

  it("filters pokemons by name", async () => {
    const result = await app.executor({
      variables: {
        filters: {
          name: "charm",
        } as PokemonFilters,
      },
      document: parse(/* GraphQL */ `
        query listPokemon($filters: ListPokemonFilters!) {
          listPokemon(filters: $filters) {
            edges {
              id
              name
            }
          }
        }
      `),
    });

    assertSingleValue(result);

    const { data, errors } = result;

    expect(errors).toBeUndefined();
    expect(data).toBeDefined();

    expect(
      data.listPokemon.edges.every((edge: PokemonModel) =>
        edge.name.startsWith("Charm")
      )
    ).toBe(true);
  });

  it("retrieves user's favorite pokemons", async () => {
    const result = await app.executor({
      variables: {
        filters: {
          favorite: true,
        } as PokemonFilters,
      },
      document: parse(/* GraphQL */ `
        query listPokemon($filters: ListPokemonFilters!) {
          listPokemon(filters: $filters) {
            edges {
              id
              name
            }
          }
        }
      `),
    });

    assertSingleValue(result);

    const { data, errors } = result;

    expect(errors).toBeUndefined();
    expect(data).toBeDefined();

    expect(data.listPokemon.edges.map((edge: PokemonModel) => edge.name))
      .toMatchInlineSnapshot(`
      [
        "Venusaur",
        "Charmander",
        "Charmeleon",
      ]
    `);
  });

  it("retrieves pokemon by id/slug", async () => {
    const searchItem = {
      id: 4,
      slug: "charmander",
    };

    const resultSlug = await app.executor({
      variables: {
        slug: searchItem.slug,
      },
      document: parse(/* GraphQL */ `
        query ($slug: String!) {
          getPokemonBySlug(slug: $slug) {
            id
            slug
            name
            types {
              id
              name
            }
            attacks {
              id
              name
              category
              damage
              type {
                id
                name
              }
            }
            classification
            commonCaptureArea
            previousEvolution {
              id
              name
              slug
            }
            nextEvolution {
              id
              name
              slug
            }
            evolutionRequirement {
              name
              amount
            }
            maxCp
            maxHp
            fleeRate
            height {
              minimum
              maximum
            }
            weight {
              minimum
              maximum
            }
            resistants {
              id
              name
            }
            votesCount
            weaknesses {
              id
              name
            }
          }
        }
      `),
    });

    assertSingleValue(resultSlug);
    const { data, errors } = resultSlug;
    expect(errors).toBeUndefined();
    expect(data).toBeDefined();
    expect(data.getPokemonBySlug.id).toBe(searchItem.id);
    expect(data.getPokemonBySlug.slug).toBe(searchItem.slug);
    sortById(data.getPokemonBySlug.attacks);
    sortById(data.getPokemonBySlug.weaknesses);
    sortById(data.getPokemonBySlug.resistants);
    sortById(data.getPokemonBySlug.types);
    expect(data.getPokemonBySlug).toMatchSnapshot();

    const resultId = await app.executor({
      variables: {
        id: searchItem.id,
      },
      document: parse(/* GraphQL */ `
        query ($id: Int!) {
          getPokemon(id: $id) {
            id
            slug
          }
        }
      `),
    });

    assertSingleValue(resultId);
    expect(resultId.errors).toBeUndefined();
    expect(resultId.data).toBeDefined();
    expect(resultId.data.getPokemon.slug).toBe(searchItem.slug);
    expect(resultId.data.getPokemon.id).toBe(searchItem.id);
  });
});
