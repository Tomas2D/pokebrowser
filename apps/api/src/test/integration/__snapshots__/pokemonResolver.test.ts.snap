// Vitest Snapshot v1

exports[`Pokemon > retrieves pokemon by id/slug 1`] = `
{
  "attacks": [
    {
      "category": "fast",
      "damage": 10,
      "id": 9,
      "name": "Ember",
      "type": {
        "id": 7,
        "name": "Fire",
      },
    },
    {
      "category": "fast",
      "damage": 6,
      "id": 10,
      "name": "Scratch",
      "type": {
        "id": 11,
        "name": "Normal",
      },
    },
    {
      "category": "special",
      "damage": 30,
      "id": 11,
      "name": "Flame Burst",
      "type": {
        "id": 7,
        "name": "Fire",
      },
    },
    {
      "category": "special",
      "damage": 25,
      "id": 12,
      "name": "Flame Charge",
      "type": {
        "id": 7,
        "name": "Fire",
      },
    },
    {
      "category": "special",
      "damage": 55,
      "id": 13,
      "name": "Flamethrower",
      "type": {
        "id": 7,
        "name": "Fire",
      },
    },
  ],
  "classification": "Lizard Pok√©mon",
  "commonCaptureArea": null,
  "evolutionRequirement": {
    "amount": 25,
    "name": "Charmander candies",
  },
  "fleeRate": 0.1,
  "height": {
    "maximum": 68,
    "minimum": 53,
  },
  "id": 4,
  "maxCp": 841,
  "maxHp": 955,
  "name": "Charmander",
  "nextEvolution": {
    "id": 5,
    "name": "Charmeleon",
    "slug": "charmeleon",
  },
  "previousEvolution": null,
  "resistants": [
    {
      "id": 1,
      "name": "Grass",
    },
    {
      "id": 6,
      "name": "Fairy",
    },
    {
      "id": 7,
      "name": "Fire",
    },
    {
      "id": 8,
      "name": "Ice",
    },
    {
      "id": 12,
      "name": "Bug",
    },
    {
      "id": 13,
      "name": "Steel",
    },
  ],
  "slug": "charmander",
  "types": [
    {
      "id": 7,
      "name": "Fire",
    },
  ],
  "votesCount": 3,
  "weaknesses": [
    {
      "id": 3,
      "name": "Water",
    },
    {
      "id": 14,
      "name": "Ground",
    },
    {
      "id": 15,
      "name": "Rock",
    },
  ],
  "weight": {
    "maximum": 9.56,
    "minimum": 7.44,
  },
}
`;
